### ISSUE TITLE
Doing a hot restart while the debugger has paused on exception hangs if before runApp

### ISSUE THREAD

**Issue by nickmeinhold** [flutter/flutter#39895]
## Steps to Reproduce

1. flutter create hello_world
2. add `throws('hello');` to main before `runApp(MyApp());`
3. in VSCode 
  - hit 'Start Debugging' 
  - under BREAKPOINTS select "All Exceptions" 
4. When debugger pauses on the Exception, hit Hot Restart
5. Dialog shows "Syncing files to device..."  with progress indicator but Hot Restart never happens 

https://gist.github.com/nickmeinhold/fe577c51a4c52a5cc8e8d97ba642fcf1

```
[✓] Flutter (Channel stable, v1.7.8+hotfix.4, on Mac OS X 10.14.6 18G95, locale en-AU)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 10.3)
[✓] iOS tools - develop for iOS devices
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.38.0)
[✓] Connected device (1 available)
```

cc @DanTup

**Comment by DanTup** [comment]
I can repro this. I suspect something in the Hot Restart code might not work correctly if `runApp` was never called, though I'm not familiar with exactly how it works (I suspect others here will though).

**Comment by nickmeinhold** [comment]
Thanks Dan!  I'm seeing in my larger projects that if an object that was created in main and passed into runApp and then later throws while the app is running, I see the same behaviour.  It happens every time in my projects but I couldn't repro in a clean project unfortunately.  I'm hoping it's the same problem as what happens here.

**Comment by nickmeinhold** [comment]
Thanks @DanTup for suggesting updating Dart-Code to 3.5.0 Alpha 3.  I tried but unfortunately still seeing the same problem.

There was a difference but the result seems to be the same.  The difference was that hitting Hot Restart with the updated Dart-Code made the red debug pop-up and the contents of the Call Stack section disappear, but otherwise everything is the same and the restart never occurs. 

Dart-Code 3.4.1, after breaking on exception then hitting Hot Restart
<img width="1439" alt="Screen Shot 2019-09-17 at 2 23 17 pm" src="https://user-images.githubusercontent.com/1059276/65012069-31a93400-d959-11e9-88f1-da96749a51e1.png">

Dart-Code 3.5.0 Alpha 3, after breaking on exception then hitting Hot Restart
<img width="1438" alt="Screen Shot 2019-09-17 at 2 15 37 pm" src="https://user-images.githubusercontent.com/1059276/65012097-4dacd580-d959-11e9-9d9a-f5b6c6ddc7e2.png">

**Comment by DanTup** [comment]
@nickmeinhold thanks for testing - re-reading your issue, I think @escamoteur's issue was slightly different (caused by the VS Code debugger being in a bad state - which I think this fixes). With yours, it does seem like the hot restart is failing to complete.

I had a quick look at this again, and I think I have a better idea of what's happening... The restart code is causing `main` to be run again, which is throwing an exception. This breaks the debugger (again), but this time the call stack does not include the correct location, so it doesn't appear to break back on the right code (you can see it shows "Paused on Exception" in the call stack pane in your screenshot, because it's paused on the exception again, as part of the restart). Here's what the call stack looked like for me:

<img width="435" alt="Screenshot 2019-09-17 at 9 50 58 am" src="https://user-images.githubusercontent.com/1078012/65026539-ab78f780-d930-11e9-80ba-bb6d59feaac5.png">

The restart never completes because it's paused on that exception. In my testing, if I *removed* the code that threw the exception (since it'll forever get stuck breaking on that exception if you don't fix it) and then clicked Resume (I had to click it twice, because the exception seemed to be thrown multiple times - maybe similar to https://github.com/dart-lang/sdk/issues/37645) then the app seemed to load correctly.

Does this happen for you?

**Comment by nickmeinhold** [comment]
Yes!  I see the same behaviour when I perform the actions you describe (including having to hit Resume twice).  

This also matches what I have been seeing in my larger projects - I have found that if I hit Resume until the debugger stops breaking, _then_ hit Hot Restart, the Hot Restart works. 

I will try again to repro the behaviour that I see in my larger projects by starting with a clean project.  The behaviour I am referring to is when an object is created in main, before runApp, and passed in to runApp and only throws later while the app is running.

**Comment by nickmeinhold** [comment]
> I will try again to repro the behaviour that I see in my larger projects by starting with a clean project. The behaviour I am referring to is when an object is created in main, before runApp, and passed in to runApp and only throws later while the app is running.

To close the 2 year old loop :-) my memory is that I was not able to reproduce the behaviour with a clean project.
